Source/WebGPU/WGSL/GlobalSorting.cpp

```diff
@@ -0,0 +1,357 @@
+        }
+        static bool equal(const Edge& a, const Edge& b)
+        {
+            return a == b;
+        }
+        static constexpr bool safeToCompareToEmptyOrDeleted = true;
+    };
+
+    Graph(size_t capacity)
+        : m_nodes(capacity)
+    {
+    }
+
+    FixedVector<Node>& nodes() { return m_nodes; }
+    Node* addNode(unsigned index, ASTNode& astNode)
+    {
+        if (m_nodeMap.find(astNode.name()) != m_nodeMap.end())
+            return nullptr;
+
+        m_nodes[index] = Node(astNode);
+        auto* node = &m_nodes[index];
+        m_nodeMap.add(astNode.name(), node);
+        return node;
+    }
+    Node* getNode(const AST::Identifier& identifier)
+    {
+        auto it = m_nodeMap.find(identifier);
+        if (it == m_nodeMap.end())
+            return nullptr;
+        return it->value;
+    }
+
+    EdgeSet& edges() { return m_edges; }
+    void addEdge(Node& source, Node& target)
+    {
+        auto result = m_edges.add(Edge(source, target));
+        Edge& edge = *result.iterator;
+        source.outgoingEdges().add(edge);
+        target.incomingEdges().add(edge);
+    }
+
+    void topologicalSort();
+
+private:
+    FixedVector<Node> m_nodes;
+    HashMap<String, Node*> m_nodeMap;
+    EdgeSet m_edges;
+};
+
+struct Empty { };
+
+template<typename ASTNode>
+class GraphBuilder : public AST::Visitor, public ContextProvider<Empty> {
+public:
+    static void visit(Graph<ASTNode>&, typename Graph<ASTNode>::Node&);
+
+    void visit(AST::Function&) override;
+    void visit(AST::Structure&) override;
+    void visit(AST::Variable&) override;
+    void visit(AST::VariableStatement&) override;
+    void visit(AST::CompoundStatement&) override;
+    void visit(AST::IdentifierExpression&) override;
+    void visit(AST::NamedTypeName&) override;
+
+private:
+    GraphBuilder(Graph<ASTNode>&, typename Graph<ASTNode>::Node&);
+
+    void introduceVariable(AST::Identifier&);
+    void readVariable(AST::Identifier&) const;
+
+    Graph<ASTNode>& m_graph;
+    typename Graph<ASTNode>::Node& m_currentNode;
+};
+
+template<typename ASTNode>
+void GraphBuilder<ASTNode>::visit(Graph<ASTNode>& graph, typename Graph<ASTNode>::Node& node)
+{
+    GraphBuilder(graph, node).visit(node.astNode());
+}
+
+template<typename ASTNode>
+GraphBuilder<ASTNode>::GraphBuilder(Graph<ASTNode>& graph, typename Graph<ASTNode>::Node& node)
+    : m_graph(graph)
+    , m_currentNode(node)
+{
+}
+
+template<typename ASTNode>
+void GraphBuilder<ASTNode>::visit(AST::Function& function)
+{
+    ContextScope functionScope(this);
+
+    for (auto& parameter : function.parameters()) {
+        AST::Visitor::visit(parameter.typeName());
+        introduceVariable(parameter.name());
+    }
+
+    AST::Visitor::visit(function.body());
+
+    if (function.maybeReturnType())
+        AST::Visitor::visit(*function.maybeReturnType());
+}
+
+template<typename ASTNode>
```

Comment:
Is it possible to include the `AST::Visitor::visit` functions in `GraphicsBuilder` with `using` statements instead of using these forwarding functions?

Reply:
Good call, updated